### ISSUE TITLE
Caching using LRU or Memcached

### ISSUE THREAD

**Issue by mogwai** [dynamoosejs/dynamoose#335]
#325 Idea: 

Would be really cool for example to be able to setup a model so that you can cache read operations locally into memory or ( if using lambda functions) a service such as MemCachier.

You could configure this in the model. 

```js
const DynamoCache = require('dynamoose-caching')

const Owner = new Schema({
  name: String,
  pets: [String],
})

const OwnerModel = dynamoose.model("Owners", Owner)

const CachingStrategy = new DynamoCache({
  strategy: DynamoCache.Strategy.MEMCACHIER,
  username: "Username",
  password: "12345678" // Super secure
  uri: "https://asdas:3000",
  operations: ["scan.all", "query"],
  timeout: 3000,
})

OwnerModel.plugin(CachingStrategy)
```

I'd be less scared to use scan operation with this ;)

**Comment by fishcharlie** [comment]
@mogwai Sounds like a great idea! At first glance this should be possible with the current implementation of plugins. I would encourage you to try to give it a shot building it and let me know your thoughts so I can improve plugin support.

I’m not sure if I’ll have time to build this anytime soon, but if you are unable to I’ll try to get around to this at some point.

**Comment by fishcharlie** [comment]
@mogwai I believe #325 is pretty stable. We might make some more changes but I believe it’s pretty stable. Although I haven’t gotten around to building my own plugin so I’m not 100% sure. So play around with it and let me know your thoughts or suggestions.

**Comment by fishcharlie** [comment]
@mogwai Just wanted to check in and see if you were able to make any progress with this. 😃

**Comment by mogwai** [comment]
Hey man I've been swamped. I'll try and start it this week

**Comment by fishcharlie** [comment]
@mogwai Any progress on this? Just want to hear your experiences with Plugin support and get some information about changes that we still need to make before getting it released.

**Comment by mogwai** [comment]
Hey Charlie, no I'm still struggling to find the time to do it. I'm catching a flight tomorrow so will have a crack at in on the plane

**Comment by mogwai** [comment]
https://github.com/mogwai/dynamoose-caching 

I've started it. 

For now because I'm no able to use the npm package directly (or atleast I don't know how to) I will use your repo as a git submodule.

**Comment by mogwai** [comment]
https://www.npmjs.com/package/dynamoose-caching

**Comment by mogwai** [comment]
@fishcharlie I'd almost finished my implementation when I last worked on the plugin so I continued where I'd left off. Considering that Q promises will be removed in favour of something else I choose to avoid using them entirely. instead, I manually override the scan function, injecting my logic to return cached values. 

Features I'd like to add some of which are already added:

- Pass specific cache options to each operation **(done)**
- Ignore the cache for a specific query `model.scan().all().exec(true)` **(done)**
- Passing an object into the plugin options that allows the user to manage the various caches and clear them.
- Having the option to make operations share a cache 
- Cache a specific query `model.scan().all().exec().cache(options)` 
- Specific a hashing function to use for the LRU Caching strategy

An example of how this works can be seen in the test that I've written. If you consider this a valid implementation I will continue to write further tests.

**Comment by fishcharlie** [comment]
@mogwai Awesome!!

So few suggestions. First, I would be hesitant to override the native Dynamoose scan function. Of course, at the end of the day, it's your choice. But I can't see how removing Q in favor of native JS promises will cause problems there. So like if there is a valid result in your cache, I do agree with resolving early, and overriding the rest of the Dynamoose function. But if there is no result in your cache for that, I feel like it should just continue using the rest of the Dynamoose function.

It sounds like one of your bigger struggles is passing options into your plugin. I think the majority of Dynamoose methods have an options parameter, which I think it would be better to use. Probably something like `options.cache.ignore` or something like that (just having a nested object), just to ensure there is no collisions with Dynamoose itself.

I have attached the line for the options for the Scan method of where you should be able to find that. We pass in the Scan instance to you. I'm PRETTY sure inside of that scan instance you should be able to get the options object. If not, let me know because I think it's something that for sure we want to expose to plugins.

https://github.com/dynamoosejs/dynamoose/pull/325/files#diff-e29ed7ad92ee67ae24506bfd53caf131R143

That would allow for syntax such as:

```js
model.scan({}, {
  cache: {
    ignore: true
  }
}).all().exec()
```

Instead of

```js
model.scan().all().exec(true)
```

Of course, just a suggestion, not required or anything like that. The goal of plugins is to give a flexible platform that people can use. I really don't want to dictate how they work. My goal is just to provide the platform so others can take advantage of it.

---

Overall, how do you feel the experience went? I'm planning on releasing the beta of Plugins in the next Dynamoose version. Do you have any feedback for that? I still think there are some big questions that need to be figured out, but I want people to start playing with it.

So any feedback you have regarding the system, improvements that can be made, etc. would be **super** helpful.

**Comment by mogwai** [comment]
I think that as long as it doesn't affect the way that other plugins work, it should be ok. If it ends up effecting them, then I will change it. 

I don't think we're discussing the same thing regarding Q promises and native promises. I'm specifically talking about the _emit call being used here: 

```
let shouldContinue = await theScan.Model._emit('model:scan', 'request:pre', {event: {scan: theScan, callback: next, scanReq: scanOneReq}, actions: {updateScanReq: function(req) {scanOneReq = req;}}}, deferredMain);
```

Would `deferredMain` be replaced with a native promise?

------

I think this isn't as çlean.

> model.scan({}, {
>   cache: {
>     ignore: true
>   }
> }).all().exec()

I think that this would be ideal and follow the style in general.

`model.scan().all().cache(options)` or `model.scan().all().ignoreCache()`

What do you think?

------

Feedback

I think that it makes sense. I would have tried to assimilate common patterns found in other libraries e.g. `chai` using `chai-http`. 

I think it all depends on how you see the Plugin API being used. What ideas do you have?

**Comment by fishcharlie** [comment]
@mogwai In terms of promises. Got it. Yes. I can see that being an issue. I'm going to do my best to ensure PR #456 will not break anything. So I would have the assumption that it will get resolved or some solution will happen before that gets merged in. Plugin support is a higher priority than removing the Q dependency.

That style sounds great! Do you think there is anything that I need to implement on the Dynamoose end to make that possible or easier for you?

In terms of what ideas we have for the plugin API. I'm honestly not really sure. I want to make it flexible enough to where everyone can bring their ideas to the platform. In PR #325 you can find a section for `Use Cases we need to support`, that includes some ideas for plugins.

**Comment by fishcharlie** [comment]
@mogwai Awesome! Yeah that is why it's going to launch in beta. Keep me posted if you have any additional feedback.

**Comment by mogwai** [comment]
Ok I've written the basic tests for this plugin. Probably best to move discussion to the plugin page. You should be able to `npm i dynamoose-caching` and use it now :)

**Comment by fishcharlie** [comment]
@mogwai Awesome! Beta support for Dynamoose will be released in the next version (#464). After that users won't have to use a forked version of Dynamoose to use your plugin.
